src/snowflake/connector/auth.py

```diff
@@ -361,80 +357,100 @@ def post_request_wrapper(self, url, headers, body):
-                         "snowflake-connector-python[secure-local-storage]")
+    def _write_temporary_credential(self, host, user, cred_type, cred):
+        if not cred:
+            logger.debug("no credential is given when try to store temporary credential")
             return
-        new_target = convert_target(host, user)
-        try:
-            keyring.set_password(new_target, user.upper(), id_token)
-        except keyring.errors.KeyringError as ke:
-            logger.debug("Could not store id_token to keyring, %s", str(ke))
-    elif IS_LINUX and store_temporary_credential:
-        write_temporary_credential_file(host, account, user, id_token)
+        if IS_MACOS or IS_WINDOWS:
+            if not installed_keyring:
+                logger.debug("Dependency 'keyring' is not installed, cannot cache id token. You might experience "
+                             "multiple authentication pop ups while using ExternalBrowser Authenticator. To avoid "
+                             "this please install keyring module using the following command : pip install "
+                             "snowflake-connector-python[secure-local-storage]")
+                return
+            try:
+                keyring.set_password(build_temporary_credential_name(host, user, cred_type), user.upper(), cred)
+            except keyring.errors.KeyringError as ke:
+                logger.debug("Could not store id_token to keyring, %s", str(ke))
+        elif IS_LINUX:
+            write_temporary_credential_file(host, build_temporary_credential_name(host, user, cred_type), cred)
+        else:
+            logger.debug("OS not support for Local Secure Storage")
+
+    def write_temporary_credentials(self, host, user, session_parameters, response):
+        if self._rest._connection.consent_cache_id_token and session_parameters.get(
+                PARAMETER_CLIENT_STORE_TEMPORARY_CREDENTIAL, False):
+            self._write_temporary_credential(host, user, ID_TOKEN, response['data'].get('idToken'))
+
+        if session_parameters.get(PARAMETER_CLIENT_REQUEST_MFA_TOKEN, False):
+            self._write_temporary_credential(host, user, MFA_TOKEN, response['data'].get('mfaToken'))
+        return
+
+
+def flush_temporary_credentials():
+    """Flush temporary credentials in memory into disk. Need to hold TEMPORARY_CREDENTIAL_LOCK."""
+    global TEMPORARY_CREDENTIAL
+    global TEMPORARY_CREDENTIAL_FILE
+    for _ in range(10):
+        if lock_temporary_credential_file():
+            break
+        time.sleep(1)
     else:
-        logger.debug("connection parameter client_store_temporary_credential not set or OS not support")
+        logger.debug("The lock file still persists. Will ignore and "
+                     "write the temporary credential file: %s",
+                     TEMPORARY_CREDENTIAL_FILE)
+    try:
+        with codecs.open(TEMPORARY_CREDENTIAL_FILE, 'w',
+                         encoding='utf-8', errors='ignore') as f:
```

Comment:
```suggestion
        with open(TEMPORARY_CREDENTIAL_FILE, 'w',
                  encoding='utf-8', errors='ignore') as f:
```
https://stackoverflow.com/a/22288895